doc/source/data/data-internals.rst

```diff
@@ -11,79 +11,164 @@ For a gentler introduction to Ray Data, see :ref:`Quickstart <data_quickstart>`.
+:meth:`~ray.data.Dataset.iter_batches`.
 
-.. image:: images/dataset-read.svg
-   :align: center
+Block size limiting
+~~~~~~~~~~~~~~~~~~~
 
-..
-  https://docs.google.com/drawings/d/15B4TB8b5xN15Q9S8-s0MjW6iIvo_PrH7JtV1fL123pU/edit
+Ray Data bounds block sizes to avoid excessive communication overhead and prevent 
+out-of-memory errors. Small blocks are good for latency and more streamed execution, 
+while large blocks reduce scheduler and communication overhead. The default range 
+attempts to make a good tradeoff for most jobs.
 
-To handle transient errors from remote datasources, Ray Data retries application-level
-exceptions.
+By default, Ray Data attempts to bound block sizes between 1 MiB and 128 MiB. If a block 
+is larger than 192 MiB (50% more than the upper bound), Ray Data splits the block into 
+smaller blocks. 
 
-For more information on loading data, see :ref:`Loading data <loading_data>`.
+Ray Data can’t split rows. So, if your dataset contains large rows (for example, large 
+images), then Ray Data can’t bound the block size.
 
-Transforming data
------------------
 
-Ray Data uses either :ref:`Ray tasks <task-key-concept>` or
-:ref:`Ray actors <actor-key-concept>` to transform blocks. By default, it uses tasks.
-Usually, transforms are fused together and with the upstream read task.
+Operators, plans, and planning
+------------------------------
 
-.. image:: images/dataset-map.svg
-   :align: center
-..
-  https://docs.google.com/drawings/d/12STHGV0meGWfdWyBlJMUgw7a-JcFPu9BwSOn5BjRw9k/edit
+Operators
+~~~~~~~~~
+
+There are two types of operators: *logical operators* and *physical operators*. Logical 
+operators are stateless objects that describe “what” to do. Physical operators are 
+stateful objects that describe “how” to do it. An example of a logical operator is 
+``ReadOp``, and an example of a physical operator is ``TaskPoolMapOperator``.
+
+Plans
+~~~~~
+
+A *logical plan* is a series of logical operators, and a *physical plan* is a series of 
+physical operators. When you call APIs like :func:`ray.data.read_images` and 
+:meth:`ray.data.Dataset.map_batches`, Ray Data produces a logical plan. When execution 
+starts, the planner generates a corresponding physical plan. 
+
+The planner
+~~~~~~~~~~~
+
+The Ray Data planner translates logical operators to one or more physical operators. For 
+example, the planner translates the ``ReadOp`` logical operator into two physical 
+operators: an ``InputDataBuffer`` and ``TaskPoolMapOperator``. Whereas the ``ReadOp``
+logical operator only describes the input data, the ``TaskPoolMapOperator`` physical 
+operator actually launches tasks to read the data.
+
+Plan optimization
+~~~~~~~~~~~~~~~~~
+
+Ray Data applies optimizations to both logical and physical plans. For example, the 
+``OperatorFusionRule`` combines a chain of physical map operators into a single map 
+operator. This prevents unnecessary serialization between map operators.
```

Comment:
add how to set optimization rules

Reply:
Don't think we expose a functioning API to configure optimization rules. `add_user_provided_logical_rules` and `add_user_provided_physical_rules` don't do anything.

I think you can configure `ray.data._internal.logical.optimizers.DEFAULT_LOGICAL_RULES`, but that seems more of a hack.